### ISSUE TITLE
Textfield that replaces another Textfield doesn't render properly

### ISSUE THREAD

**Issue by aaugustin** [tleunen/react-mdl#197]
I encontered an issue when rendering a component that dynamically decides which fields are visible.

Here's a small example that exhibits the issue. It's a form with two fields, only one of which is visible at a given time. To keep things simple, I put a switch to select which field is visible. In my actual use case, this depends on the values of previous fields.

```javascript
import {Switch, Textfield, Button} from 'react-mdl';

const TestComponent = React.createClass({

    getInitialState() {
        return {
            showFirst: true,
        }
    },

    onChangeSelector(event) {
        this.setState({showFirst: !this.state.showFirst});
    },

    onChangeFirst(event) {
        this.setState({first: event.target.value});
    },

    onChangeSecond(event) {
        this.setState({second: event.target.value});
    },

    render() {
        return (
            <div>
                <Switch checked={this.state.showFirst} onChange={this.onChangeSelector}>Show first field</Switch>
                {
                    this.state.showFirst
                    ? <Textfield label="first" value={this.state.first} onChange={this.onChangeFirst} />
                    : <Textfield label="second" value={this.state.second} onChange={this.onChangeSecond} />
                }
                <pre>state = {JSON.stringify(this.state)}</pre>
            </div>
        )
    },

});
```

Initially the first field is visible. After typing "abc" in the first field, I get this, which looks good:

<img width="379" alt="screen shot 2016-01-30 at 19 23 58" src="https://cloud.githubusercontent.com/assets/788910/12697588/1aceb66a-c788-11e5-9237-f2df7ebfeb35.png">

Then I flick the switch. After typing "def" in the second field, I get this, which looks less good:

<img width="508" alt="screen shot 2016-01-30 at 19 24 11" src="https://cloud.githubusercontent.com/assets/788910/12697593/3301cdb2-c788-11e5-9bfd-ef0bc6b317f2.png">
 
I believe this is the symptom of a desynchronization between the React and MDL components lifecycles. I'll see if I can track this down further.

**Comment by aaugustin** [comment]
Giving the two fields different keys i.e. `<Textfield key="first/second" ... />` makes the problem go away. This suggests that the issue arises when React [reconciles the two Textfields](https://facebook.github.io/react/docs/reconciliation.html#custom-components).

It may have something to do with the fact that `MDLComponent` implements `componentDidMount` / `componentWillUnmont` but not `componentWillUpdate` / `componentDidUpdate`.

**Comment by tleunen** [comment]
Indeed, `Textfield` might have an issue with the internal state. But the doc you mentioned specifically says to specify a key to let React know the element is different (which is the case in your example).

**Comment by aaugustin** [comment]
Yes, I couldn't find a convincing way to handle this in Textfield. Perhaps it's best to treat this as a known limitation and to document the workaround. My use case may not be very common.

**Comment by aaugustin** [comment]
For my own use I have a wrapper that sets `key={props.name}` but I don't think this solution can be generalized because:

- not everyone will set unique names on fields because they aren't necessary if the form submission is handled in JavaScript based on the React state
- it's likely to have unintended consequences; it may do more harm than good

**Comment by tleunen** [comment]
I'll keep the ticket open to investigate on this.

But for your usecase, please use a key prop. Not everyone will swap textfield based on a condition, so for most usecases, it behaves correctly.

I ported your example in a codepen to better visualize the issue: http://codepen.io/anon/pen/EPRLww

**Comment by tleunen** [comment]
The issue is because we set the `key` prop on the Textfield. Because of that, the input changed, but not the whole element, so it's never unmounted / re-mounted.

I'll try to find a fix for the 1.4 release...
